url64b10c0106_schlabberdog_blocks_tgz-pJ8-com_github_users_schlabberdog_blocks_ui_CliUIJ8/src/com/github/users/schlabberdog/blocks/solver/Solver.java

```diff
@@ -0,0 +1,264 @@
+import com.github.users.schlabberdog.blocks.board.moves.IMove;
+
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Stack;
+
+public class Solver {
+    private final Board board;
+    private final ISolutionChecker checker;
+
+	private ISolverDelegate delegate;
+
+    private long checkCount;
+    private int solutionCount;
+    private int solutionImprovedCount;
+    private int bestPathLength;
+	private int worstStack;
+
+	private boolean avoidWorseStacks = true;
+	private int stackDepthLimit = 0;
+
+    /* In dem Array bewahren wir uns alle Schritte auf, die wir gemacht haben */
+    private LeveledSteps steps = new LeveledSteps();
+
+    private Stack<Backtrack> btStack = new Stack<Backtrack>();
+
+    public Solver(Board board, ISolutionChecker checker) {
+        this.board = board;
+        this.checker = checker;
+    }
+
+    public synchronized void solve() {
+	    solverStarted();
+
+        steps.clear();
+        btStack.clear();
+        checkCount = 0;
+        solutionCount = 0;
+        solutionImprovedCount = -1;
+	    worstStack = 0;
+	    bestPathLength = Integer.MAX_VALUE;
+        //ausgangssituation auf level 0 (ungschlagbar)
+        steps.pushOnLevel(board.getBoardHash(),0);
+        //um das ganze anzustoßen müssen wir zuerst die alternativen für Schritt 0 aufstellen
+        goDeeper();
+
+        while(true) {
+	        //solange wir noch alternativen haben...
+	        if (!goAnywhere())
+		        break;
+        }
+
+	    solverDone();
+    }
+
+    private boolean goAnywhere() {
+
+        if(btStack.isEmpty())
+            return false;
+        // vielleicht haben wir auf dieser ebene noch alternativen?
+        if(btStack.peek().alternatives.size() > 0) {
+
+            goRight();
+
+        }
+        //nein haben wir nicht, d.h. wir müssen eins zurück und nach rechts
+        else {
+            goUp();
+
+        }
+		return true;
+    }
+
+    private void goRight() {
+
+        //nach rechts gehen bedeutet:
+        Backtrack cur = btStack.peek();
+        // 1. board zum anfangswert dieses stack-levels resetten
+
+        board.applySave(cur.initialState);
+
+        // 2. alternative anwenden und verwerfen
+        IMove m = cur.alternatives.get(0);
+        cur.alternatives.remove(0);
+
+        board.applyMove(m);
+
+	    cur.selected = m;
+
+        goDeeper();
+    }
+
+	private void checkSolution() {
+		//prüfen wir, ob das was wir haben eine lösung ist
+		checkCount++;
+		if(checker.checkBoard(board)) {
+			solutionCount++;
+
+			int pathLength = (btStack.size() - 1); //-1 ist notwendig weil die ausgangsposition auch auf dem stack liegt
+
+			//uns interessieren nur lösungen, die besser sind als bereits bekannte
+			if(pathLength < bestPathLength) {
+				solutionImprovedCount++;
+				bestPathLength = pathLength;
+
+				solutionImproved(pathLength);
+			}
+		}
+	}
+
+
+    private void goDeeper() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.